### ISSUE TITLE
MetaData

### ISSUE THREAD

**Issue by gordonwatts** [ssl-hep/ServiceX#664]
# Story
_ As a developer on ServiceX, I want to know all the types of metadata that we might want to use so I can figure out how to implement an API._

# Acceptance Criteria
1. We have a good list of metadata of different types, for ntuple files, xAOD files, CMS nanoAOD files, CMS miniAOD files.

**Comment by ponyisi** [comment]
The chief metadata of interest to analysis, I think, is a sum of (pre-selection) event weights in Monte Carlo. It might also be useful to return number of pre-selection rows.

This raises what is I suspect an important question - do we need to allow the user to specify the required metadata and how to compute it? e.g. if I just have some random ntuple that's not part of a standard ATLAS format, and I compute the sum of weights via some specific formula (e.g. nominal weight * PDF variation 1, nominal * PDF variation 2, ...) - do we need an API to specify this query? For xAOD/nanoAOD I imagine that this will be quite standard and it's not necessary.

**Comment by gordonwatts** [comment]
From the xAOD POV:
* The CutFlow table info that can be found so you can get at the original files
* Isn't there some sort of lumi information stored (so you can drop files and your denominator isn't killed?)

**Comment by alexander-held** [comment]
I will start with a question: are we talking here exclusively about metadata made available to the user in the ServiceX output, or also about metadata available at runtime for the transformer to use?

For custom user-created ntuples I think it is difficult to say anything generic: users can in principle store anything they want in pretty much any format they want.

I am adding a few thoughts from an ATLAS perspective (we should hear from CMS too). A popular format at the moment are AnalysisTop (DAOD->ntuple) outputs, where two ways exist to extract information about the total (weighted) sum of events:
- histogram bin look-up (need to sum histograms over all files and then look at bin content),
- metadata tree branch (sum values stored in branches of specific tree across all files).
I hope I am remembering those correctly. As far as I can see at the moment, the AnalysisTop successor (TopCPToolkit) also uses the histogram method.

I think the most important information is that weighted or unweighted sum of events. For samples with built-in weight-based variations (e.g. PDF variations), many of such sums may exist (one per variation). A bit of a niche use case for weight-based variations is also extracting the list of strings of variation names, which would sit in a special metadata branch. That might be convenient to be able to access too.

Another useful piece of metadata would be the filename and potentially (in case of rucio) also a container name. At the moment the names are obfuscated a bit with `/` -> `:` replacement. I think it is fairly common to implement DSID-specific behavior, so one could also think about providing that directly (not hard to parse from filename otherwise).

**Comment by gordonwatts** [comment]
Also consider the difference between "metadata" that is downstream and "metadata" that is part of the file:

* Extracting information from other than CollectionTree (e.g. a second tree)
* Extracting metadata from the original CollectionTree, for example, sum of all weights multiplied by a pdf (per event, and then sum)

These are two types of queries but both are called metadata.

From a conversation with @ponyisi

**Comment by ponyisi** [comment]
After further discussion with @gordonwatts we agreed that
* the metadata configuration and the returned values should be opaque to the ServiceX backend (for the client libraries we may need to have codegen-specific helper methods);
* the current transformer configuration methods (qastle and the uproot-raw JSON) can support the metadata as part of the primary request, so no additional ServiceX API is needed to get the information into the transformers.

We will need to discuss internal ServiceX API handling of the metadata output from the transformers, but this should be a relatively simple path.

**Comment by gordonwatts** [comment]
An example of xAOD metadata [can be found here](https://twiki.cern.ch/twiki/bin/viewauth/AtlasProtected/AthAnalysisBase#How_to_print_the_sum_of_weights).

**Comment by ponyisi** [comment]
So this kind of fell through the cracks. For uproot-raw I feel all intended use cases have been addressed. Perhaps @alexander-held, @gordonwatts have some examples from the 200 Gbps test?

**Comment by alexander-held** [comment]
For 200 Gbps so far I've not hit any particular metadata issues, largely due to the fact that we stripped out a lot of the more physics-relevant pieces. When we add back more physics again that might change.
